CompactVector.java

```diff
@@ -11,32 +11,90 @@ public CompactVector() {
 
 
 
-    public void Read() {
-
+    public int Read(int index, int oldElem) {
+    	
+    	// check bounds
+    	if(index > this.array.length) 
+    	{
+    		// return value indicating read failed? 
+    		return -1;
+    	}
+    	// check transaction status
+    	if(this.array[index].transaction.txnStatus.get().equals(TxnStatus.active)) 
+    	{
+    		// helping must occur for this transaction, so abort?
+    		// reference page 6 section 6 paragraph 2 
+    		return -1;
+    	}
+    	if(this.array[index].transaction.txnStatus.get().equals(TxnStatus.committed)) 
+    	{
+    		//committed is the only status to return new value
+    		return this.array[index].newValue;
+    	}
+    		
+    	return this.array[index].oldValue;
     }
 
 
-    public void Write() {
-
+    public void Write(int index, int newval) {
+    	
+    	if(index < this.array.length) 
+    	{
+    		this.array[index].oldValue = this.array[index].newValue;
+    		this.array[index].newValue = newval;
+    	}
+    	// Do we need to put anything in the return value to signify success?
+    		
+    	
     }
 
 
-    public void PushBack() {
-
+    public void PushBack( int newval) {
+    	
+    	int len = array.length-1;
+    	this.array[len].oldValue = this.array[len].newValue;
+    	this.array[len].newValue = newval;
```

Comment:
I was reading the c++ definition of push_back and we may need to change this. Here is the c++ definition: 

"Adds a new element at the end of the vector, after its current last element. The content of val is copied (or moved) to the new element.

This effectively increases the container size by one, which causes an automatic reallocation of the allocated storage space if -and only if- the new vector size surpasses the current vector capacity."

So its almost like we choose the index after the furthest allocated index, and if array.length-1 is already taken we need to increase capacity?